slick/src/main/scala/slick/util/AsyncExecutor.scala

```diff
@@ -22,16 +22,19 @@ object AsyncExecutor extends Logging {
     * @param numThreads The number of threads in the pool.
     * @param queueSize The size of the job queue, 0 for direct hand-off or -1 for unlimited size. */
   def apply(name: String, numThreads: Int, queueSize: Int): AsyncExecutor = {
-    val queue = queueSize match {
-      case 0 => new SynchronousQueue[Runnable]
-      case -1 => new LinkedBlockingQueue[Runnable]
-      case n => new ArrayBlockingQueue[Runnable](n)
-    }
-    val tf = new DaemonThreadFactory(name + "-")
-    val executor = new ThreadPoolExecutor(numThreads, numThreads, 1, TimeUnit.MINUTES, queue, tf)
     new AsyncExecutor {
-      val executionContext = ExecutionContext.fromExecutorService(executor, loggingReporter)
-      def close(): Unit = executor.shutdownNow()
+      private[this] var executor: ThreadPoolExecutor = _
```

Comment:
You need to enforce a happens-before relationship externally anyway. Otherwise there's no guarantee that you are not calling close() while the executor is still needed. Hm, maybe we should change that, too.